### PR TITLE
Refactor to transform graph to use strings before passing to Python

### PR DESCRIPTION
**Pull request by r-barnes**
Also adding type annotations for the pybind11 extension and calculating `total_area` before we get to Python so that Python's view of the data is never invalid.

### CODE REVIEW

**Comment by r-barnes** [comment]
This differs from `main` if the regularizers are enabled on both. I'm not sure if that means there's a mistake on main or a mistake on this branch :-(

**Review by AlexLipp — CHANGES_REQUESTED** [review]
Can we update the geospatial output (currently `labels.tif`) to map the sample string to the subcatchment areas? This current change means the the `get_unique_upstream_areas` function is not working correctly as it relied on a particular ordering of then nodes (which was in any case undesirable). Mapping areas by the strings makes this easier. Not sure if `.tif` allows for strings though so this might need to be a different raster file format. Alternatively, a key could be made that maps the node numbers used in `labels.tif` onto sample numbers.

File: src/faster-unmixer.cpp [hunk 1]
```diff
@@ -131,6 +222,9 @@ std::pair<std::vector<SampleNode>, adjacency_graph_t> faster_unmixer(const std::
     }
   }
 
+  // Save regions output
+  sample_label.saveGDAL("labels.tif");
+
```
[the remark this answers is not in the document]
Comment (r-barnes) [reply src/faster-unmixer.cpp]: I believe I've fixed this now.

**Review by AlexLipp — APPROVED** [review]
I checked the results of these changes locally and the matching between areas and sample sites is fixed, producing sensible outputs again.

**Pull request merged**